### PR TITLE
[ActiveRecord] Support nested connection pinning

### PR DESCRIPTION
**Pull request by palkan**
### Motivation / Background

Follow-up for #50999.

See also [this discussion](https://github.com/rails/rails/pull/50999#issuecomment-2150906497).

**tl;dr** Current implementation makes it much harder to use outer transactions w/ pinning (prev. `lock_thread=true`) with Rails test fixtures (the approach widely used via TestProf's helpers, such as `before_all` and `let_it_be`).

### Detail

This PR makes it possible to nest `#pin_connection!` calls instead of raising if there is already a pinned connection. Now, it's safe to wrap code with `#pin_connection!` / `#unpin_connection!` without checking first if there is already a pinned connection or not.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 1]
```diff
@@ -311,9 +312,9 @@ def connection
       end
 
       def pin_connection!(lock_thread) # :nodoc:
-        raise "There is already a pinned connection" if @pinned_connection
```
Comment (palkan) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:314]: One thing I want to discuss is whether we want to check that the `lock_thread` value is consistent or not? I mean, how to treat the following scenarios:

```ruby
# 1

pin_connection!(true) #=> locked
pin_connection!(false) #=> still locked
unpin_conenction!
#=> locked 'till the end
...

# 2
pin_connection!(false) #=> unlocked
pin_connection!(true) #=> not it's locked
unpin_connection!
#=> still locked here! — do we want to be unlocked here?
...
```

So, right we keep the lock once it was set 'till the very final `#unpin_connection!`.

I can add a stack to keep locked values and restore in the `#unpin_connection!` accordingly, but.. do we need such complication?

Alternatively, we can treat non-locked pinned connections as _non-joinable_ with locked ones, i.e.:

```ruby
pin_connection!(false)
pin_connection!(false) # this is fine
pin_connection!(true) # !!! raises "Can not lock already pinned connection"

```

Or we can just ignore this scenario for now.

@byroot Thoughts?

File: activerecord/test/cases/connection_pool_test.rb [hunk 2]
```diff
@@ -934,6 +934,24 @@ def test_unpin_connection_returns_whether_transaction_has_been_rolledback
...
+        @pool.unpin_connection!
+        assert_instance_of RealTransaction, @pool.lease_connection.current_transaction
+        @pool.unpin_connection!
+        assert_instance_of NullTransaction, @pool.lease_connection.current_transaction
+
+        e = assert_raises(RuntimeError) do
+          @pool.unpin_connection!
+        end
+
+        assert_match(/There isn't a pinned connection/, e.message)
```
Comment (casperisfine) [inline activerecord/test/cases/connection_pool_test.rb:952]: nitpick:
```suggestion
        assert_raises(RuntimeError, match: /There isn't a pinned connection/) do
          @pool.unpin_connection!
        end
```

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb (hunk 1 above)
Comment (casperisfine) [reply activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:314]: Yeha, that's the kind of weirdness I feared. That said, I think we can consider that code that don't ask for the lock, is likely OK with the lock existing. Meaning we can treat the lock as an upgrade:

  - Calling `pin_connection(true)` should add the lock if not already there
  - Calling `pin_connection(false)` should not remove the lock if there is one.
  - Only the final `unpin_connection` should remove the lock.

Comment (byroot) [reply activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:314]: Also I don't think we really want to create nested transactions. We should only call `begin_transaction` if `depth == 1`.

**Comment by byroot** [comment]
Alright, I reworked your PR a little bit, let me know if it still works for you and if so I'll merge and backport.

**Comment by palkan** [comment]
>  let me know if it still works for you and if so I'll merge and backport.

Unfortunately, without nested transactions that wouldn’t work for me.

I need to preserver the current transactional behavior of test fixtures independently of if there is an outer pinning or not. That’s the whole point of before_all—having some shared setup between tests but still ignoring changes made in individual examples.

So, I’d suggest bringing back transactions.

**Comment by byroot** [comment]
> I need to preserver the current transactional behavior of test fixtures independently of if there is an outer pinning or not.

Hum, I see how I missed that only opening one transaction breaks transactional fixtures is your gem opens it before 🤦 .

But you don't need a transaction right, only the pinning? If so we should go with what you suggested, decouple both so you can pin the connection without opening a transaction.

**Comment by palkan** [comment]
> But you don't need a transaction right, only the pinning? 

I do need a transaction, too. This is how it works:

```rb
class SomeTest
  # clean database
  before_all do  # begin;
    # prepare some data
    # insert into ....
  end

  # savepoint
  test "one" do
     # select <data created within before_all>
     # insert/update — changes do not escape the example
  end
  # release

  # savepoint
  test "two" do
    # ditto
  end
  # release

  # exit the before_all context
  # rollback; 

  # database is clean again
end
```

Thus, we need:

- nested transitions to guarantee data/changes isolation and avoid state leakage (both in the example and example group contexts)

- pinning to avoid problems with threaded code (like AJ async or whatever)

Thus, **both test_fixtures and before_all need transactions and pinning**.

My original suggestion was to extract transactions from pinning and bring them back to test_fixtures (and manually manage transactions in before_all). This was probably a premature optimization suggestion based on the idea that _someone else_ might want to use pinning without transactions. But, we don't have such known use cases.

Maybe, to minimize changes, we can add `transactional: true` option to `#pin_connection!` / `#unpin_connection!`? This way, we don't need to touch test_fixtures anymore and add a bit of flexibility to the API. Though, it could be done in the future releases if there is a demand for that.

**Comment by byroot** [comment]
Alright, I think I get it now, I'll restore the transaction.

**Comment by byroot** [comment]
Should be good now, if you confirm it's working for you I'll merge.

**Comment by palkan** [comment]
Works perfectly (and looks nice and simple): https://github.com/test-prof/test-prof/blob/2209fd7d863c884649b833eaf22b288cd3fe499e/lib/test_prof/before_all/adapters/active_record.rb#L11  

Thanks!

**Pull request merged**